types/src/client.rs

```diff
@@ -157,17 +167,26 @@ where
 	Notif: DeserializeOwned,
 {
 	/// Returns the next notification from the stream.
-	/// This may return `Ok(None)` if the subscription has been terminated,
-	/// may happen if the channel becomes full or is dropped.
-	pub async fn next(&mut self) -> Result<Option<Notif>, Error> {
-		match self.notifs_rx.next().await {
-			Some(n) => match serde_json::from_value::<NotifResponse<Notif>>(n) {
-				Ok(NotifResponse::Ok(parsed)) => Ok(Some(parsed)),
-				Ok(NotifResponse::Err(e)) => Err(Error::SubscriptionClosed(e)),
-				Err(e) => Err(Error::ParseError(e)),
-			},
-			None => Ok(None),
-		}
+	/// This may return `None` if the subscription has been terminated,
+	/// which may happen if the channel becomes full or is dropped.
+	pub async fn next(&mut self) -> Option<Result<Notif, Error>> {
+		StreamExt::next(self).await
+	}
+}
+
+impl<Notif> Stream for Subscription<Notif>
+where
+	Notif: DeserializeOwned,
+{
+	type Item = Result<Notif, Error>;
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> task::Poll<Option<Self::Item>> {
+		let n = futures_util::ready!(self.notifs_rx.poll_next_unpin(cx));
+		let res = n.and_then(|n| match serde_json::from_value::<NotifResponse<Notif>>(n) {
```

Comment:
does `map` work here?

I don't mind but clippy doesn't like `and_then` here...

Reply:
Ooh good point! I thought I needed `and_then` but then didn't realise that I then surrounded all of the result arms in `Some` :)